lib/buildFactory.js

```diff
@@ -3,8 +3,63 @@
 const BaseFactory = require('./baseFactory');
 const Build = require('./build');
 const imageParser = require('docker-parse-image');
+const buildClusterAnnotation = 'screwdriver.cd/buildCluster';
+const hoek = require('hoek');
 let instance;
 
+/**
+ * Get build cluster name based on annotations
+ * @method getBuildCluster
+ * @param  {Object}         config
+ * @param  {Object}         config.annotations              Annoations for the job
+ * @param  {PipelineModel}  config.pipeline                 Pipeline model
+ * @return {String}        Build cluster name
+ */
+async function getBuildClusterName({ annotations, pipeline }) {
+    const buildClusterName = annotations[buildClusterAnnotation] || '';
+    // Lazy load factory dependency to prevent circular dependency issues
+    // https://nodejs.org/api/modules.html#modules_cycles
+    // eslint-disable-next-line global-require
+    const BuildClusterFactory = require('./buildClusterFactory');
+    const buildClusterFactory = BuildClusterFactory.getInstance();
+
+    if (!buildClusterName) {
+        return buildClusterFactory.list({
+            managedByScrewdriver: true,
+            isActive: true
+        }).then((buildClusters) => {
+            if (buildClusters.length === 0) {
+                return '';
+            }
+
+            // Pick a random screwdriver build cluster
+            const picked = buildClusters[Math.floor(Math.random() * buildClusters.length)];
+
+            return picked.name;
+        });
+    }
+
+    return buildClusterFactory.get({
+        name: buildClusterName
+    }).then((buildCluster) => {
+        if (!buildCluster) {
+            throw new Error('Cluster specified in screwdriver.cd/buildCluster does not exist.');
+        }
+
+        // Check if this pipeline is authorized to use the build cluster
+        // pipeline name example: screwdriver-cd/ui
+        const regex = /^([^/]+)\/.*/;
+        const matched = pipeline.name.match(regex);
+        const org = matched ? matched[1] : '';
+
+        if (!buildCluster.scmOrganizations.includes(org)) {
```

Comment:
should we save `scmOrganizations` as `<scmContext>:<org>` (ie: `partner.git:screwdriver`) or do some check for the `scmContext`?